### PR TITLE
rpi-libcamera-apps: add new recipe

### PR DESCRIPTION
**Pull request by tewarid**
Added a recipe for Raspberry Pi's libcamera-apps. Tested with poky/oe master. I haven't built or tested optional dependencies to qt, tensorflow lite, and opencv. These should be considered work in progress, for those who require these dependencies to enhance and test, and submit future PRs.

### CODE REVIEW

File: dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend [hunk 1]
```diff
@@ -1,2 +1,3 @@
-PACKAGECONFIG[raspberrypi] = "-Dpipelines=raspberrypi"
+SRCREV = "ed591e705c451d0ce14988ae96829a31a2ae2f9a"
```
Comment (agherzan) [inline dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend:1]: This should be split into another commit with the rationale of its change.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    ${@bb.utils.contains("DISTRO_FEATURES", "x11", "-DENABLE_X11=1", "-DENABLE_X11=0", d)} \
+    ${@bb.utils.contains("MACHINE", "raspberrypi3", "-DENABLE_COMPILE_FLAGS_FOR_TARGET=armv8-neon", "", d)} \
+    ${@bb.utils.contains("MACHINE", "raspberrypi4", "-DENABLE_COMPILE_FLAGS_FOR_TARGET=armv8-neon", "", d)} \
+"
+
+PACKAGECONFIG[qt] = "-DENABLE_QT=1"
+PACKAGECONFIG[opencv] = "-DENABLE_OPENCV=1"
+PACKAGECONFIG[tensorflow-lite] = "-DENABLE_TFLITE=1"
+
+FILES:${PN} = "${bindir} \
```
Comment (agherzan) [inline recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:37]: Are these needed?

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+LICENSE = "BSD-2-Clause"
+LIC_FILES_CHKSUM = "file://license.txt;md5=a0013d1b383d72ba4bdc5b750e7d1d77"
+
+S = "${WORKDIR}/git"
+
+COMPATIBLE_MACHINE = "^rpi$"
+
+DEPENDS = "libcamera libexif jpeg tiff libpng boost"
+RDEPENDS:${PN} = "python3-core libcamera libexif jpeg tiff libpng"
```
Comment (agherzan) [inline recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:20]: I would expect the build system to pick up most of these. Any reason why that is not the case here?

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+copy the functionality of the existing \"raspicam\" apps."
+HOMEPAGE = "https://github.com/raspberrypi/libcamera-apps"
+SECTION = "console/utils"
+
+SRC_URI = "\
+    git://github.com/raspberrypi/libcamera-apps.git;protocol=https;branch=main \
+"
+SRCREV = "22a52590c33a813743b4e6337478c208201c77b1"
+
+LICENSE = "BSD-2-Clause"
```
Comment (agherzan) [inline recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:12]: Small NIT: I usually have the license before the SRC_URI and after the main SUMMARY-SECTION block.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+EXTRA_OECMAKE = "\
+    -DCMAKE_BUILD_TYPE=Release \
+    -DBoost_INCLUDE_DIR=${STAGING_INCDIR} \
+    -DCMAKE_LIBRARY_PATH=${STAGING_LIBDIR} \
+    ${@bb.utils.contains("DISTRO_FEATURES", "x11", "-DENABLE_X11=1", "-DENABLE_X11=0", d)} \
+    ${@bb.utils.contains("MACHINE", "raspberrypi3", "-DENABLE_COMPILE_FLAGS_FOR_TARGET=armv8-neon", "", d)} \
+    ${@bb.utils.contains("MACHINE", "raspberrypi4", "-DENABLE_COMPILE_FLAGS_FOR_TARGET=armv8-neon", "", d)} \
+"
+
+PACKAGECONFIG[qt] = "-DENABLE_QT=1"
```
Comment (agherzan) [inline recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:33]: Should we have a default value for PACKAGECONFIG?

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+PACKAGECONFIG[opencv] = "-DENABLE_OPENCV=1"
```
Comment (agherzan) [inline recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:34]: We should also define the disable option.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+PACKAGECONFIG[tensorflow-lite] = "-DENABLE_TFLITE=1"
```
Comment (agherzan) [inline recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:35]: We should also define the disable option.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+DEPENDS = "libcamera libexif jpeg tiff libpng boost"
+RDEPENDS:${PN} = "python3-core libcamera libexif jpeg tiff libpng"
+
+inherit cmake pkgconfig
+
+EXTRA_OECMAKE = "\
+    -DCMAKE_BUILD_TYPE=Release \
+    -DBoost_INCLUDE_DIR=${STAGING_INCDIR} \
+    -DCMAKE_LIBRARY_PATH=${STAGING_LIBDIR} \
+    ${@bb.utils.contains("DISTRO_FEATURES", "x11", "-DENABLE_X11=1", "-DENABLE_X11=0", d)} \
```
Comment (agherzan) [inline recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:28]: Would it make sense to have this defined as a PACKAGECONFIG?

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    ${@bb.utils.contains("MACHINE", "raspberrypi3", "-DENABLE_COMPILE_FLAGS_FOR_TARGET=armv8-neon", "", d)} \
```
Comment (agherzan) [inline recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:29]: This would be a bit cleaner to use append overrides on the respective machines.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+"
+
+PACKAGECONFIG[qt] = "-DENABLE_QT=1"
+PACKAGECONFIG[opencv] = "-DENABLE_OPENCV=1"
+PACKAGECONFIG[tensorflow-lite] = "-DENABLE_TFLITE=1"
+
+FILES:${PN} = "${bindir} \
+    ${libdir} \
+"
+SOLIBS = ".so"
```
Comment (agherzan) [inline recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:40]: I would add a comment to explain why this is needed. Probably unversioned libraries.

**Review by agherzan — CHANGES_REQUESTED** [review]

File: dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend (hunk 1 above)
Comment (tewarid) [reply dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend:1]: Will do - we need to build a version of libcamera known to work with libcamera-apps and the kernel. In the past, I've run into issue where libcamera is not compatible with kernel and libcamera-apps.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 2 above)
Comment (tewarid) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:37]: Are you referring to PACKAGECONFIG for opencv and tensorflow lite?

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 3 above)
Comment (tewarid) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:20]: Are you referring to DEPENDS, RDEPENDS, or both? DEPENDS is needed otherwise I get build errors. RDEPENDS I am not so sure.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 5 above)
Comment (tewarid) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:33]: Why is that needed?

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 8 above)
Comment (tewarid) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:28]: Not sure - will look into it.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 9 above)
Comment (tewarid) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:29]: Could you give an example please.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 10 above)
Comment (tewarid) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:40]: Will try without it to see if it works. In the past, I had to manually create symlinks in do_install:append task.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 3 above)
Comment (agherzan) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:20]: I was only aiming the comment at RDEPENDS.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 2 above)
Comment (agherzan) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:37]: No, I'm referring to `FILES:${PN}` definition. Wouldn't bindir be part of the defaults?

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 9 above)
Comment (agherzan) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:29]: `EXTRA_OECMAKE:append:raspberrypi3 = " foo"`

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 5 above)
Comment (agherzan) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:33]: For example, if you turn `x11` into a `PACKAGECONFIG`, you could default to enabling it for the relevant distro feature.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 6 above)
Comment (agherzan) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:34]: Yes, it supports disabling and optional dependencies. https://docs.yoctoproject.org/ref-manual/variables.html?highlight=packageconfig#term-PACKAGECONFIG

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 2 above)
Comment (tewarid) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:37]: Not sure - will test without it.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 3 above)
Comment (tewarid) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:20]: Will check if build works without RDEPENDS.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 2 above)
Comment (tewarid) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb]: Get the following build error without bindir

```log
ERROR: rpi-libcamera-apps-git-r0 do_package: QA Issue: rpi-libcamera-apps: Files/directories were installed but not shipped in any package:
  /usr/bin/libcamera-raw
  /usr/bin/camera-bug-report
  /usr/bin/libcamera-vid
  /usr/bin/libcamera-jpeg
  /usr/bin/libcamera-still
  /usr/bin/libcamera-hello
```

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 10 above)
Comment (tewarid) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb]: Get build error without

```bash
SOLIBS = ".so"
FILES_SOLIBSDEV = ""
```


```log
ERROR: rpi-libcamera-apps-git-r0 do_package_qa: QA Issue: -dev package rpi-libcamera-apps-dev contains non-symlink .so '/usr/lib/libimages.so'
-dev package rpi-libcamera-apps-dev contains non-symlink .so '/usr/lib/libcamera_app.so'
-dev package rpi-libcamera-apps-dev contains non-symlink .so '/usr/lib/libpreview.so'
-dev package rpi-libcamera-apps-dev contains non-symlink .so '/usr/lib/liboutputs.so'
-dev package rpi-libcamera-apps-dev contains non-symlink .so '/usr/lib/libencoders.so'
-dev package rpi-libcamera-apps-dev contains non-symlink .so '/usr/lib/libpost_processing_stages.so' [dev-elf]
```

I'll leave it in and add a comment that it is needed due to unversioned libraries.

File: dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend (hunk 1 above)
Comment (tewarid) [reply dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend:1]: I'lll try to build libcamera-apps with upstream version of libcamera. I suspect the build won't work but let's see.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 3 above)
Comment (tewarid) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb]: Build fails due to QA error

```log
ERROR: rpi-libcamera-apps-git-r0 do_package_qa: QA Issue: /usr/bin/camera-bug-report contained in package rpi-libcamera-apps requires /usr/bin/python3, but no providers found in RDEPENDS:rpi-libcamera-apps? [file-rdeps]
```

I can leave `RDEPENDS:${PN} = "python3-core"` or eliminate /usr/bin/camera-bug-report. Since not all systems have python, the latter sounds like a better option.

File: dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend (hunk 1 above)
Comment (tewarid) [reply dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend]: As predicted, the build fails. We could chase a version of libcamera-apps that does build, or submit a patch upstream to bump libcamera version. I'm inclined to do the latter.

```log
| In file included from /home/yoctouser/amora/build/tmp/work/arm1176jzfshf-vfp-poky-linux-musleabi/rpi-libcamera-apps/git-r0/git/encoder/null_encoder.cpp:12:
| In file included from /home/yoctouser/amora/build/tmp/work/arm1176jzfshf-vfp-poky-linux-musleabi/rpi-libcamera-apps/git-r0/git/encoder/null_encoder.hpp:15:
| In file included from /home/yoctouser/amora/build/tmp/work/arm1176jzfshf-vfp-poky-linux-musleabi/rpi-libcamera-apps/git-r0/git/core/video_options.hpp:14:
| In file included from /home/yoctouser/amora/build/tmp/work/arm1176jzfshf-vfp-poky-linux-musleabi/rpi-libcamera-apps/git-r0/git/core/options.hpp:21:
| In file included from /home/yoctouser/amora/build/tmp/work/arm1176jzfshf-vfp-poky-linux-musleabi/rpi-libcamera-apps/git-r0/git/core/logging.hpp:1:
| In file included from /home/yoctouser/amora/build/tmp/work/arm1176jzfshf-vfp-poky-linux-musleabi/rpi-libcamera-apps/git-r0/git/core/libcamera_app.hpp:34:
| /home/yoctouser/amora/build/tmp/work/arm1176jzfshf-vfp-poky-linux-musleabi/rpi-libcamera-apps/git-r0/git/core/stream_info.hpp:11:10: fatal error: 'libcamera/color_space.h' file not found
| #include <libcamera/color_space.h>
|          ^~~~~~~~~~~~~~~~~~~~~~~~~
| 1 error generated.
```

Comment (tewarid) [reply dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend]: Submitted PR upstream - https://github.com/openembedded/meta-openembedded/pull/619.

Comment (tewarid) [reply dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend]: We'll have to hold off on merging this PR until upstream PR is completed.

Comment (agherzan) [reply dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend]: Sounds good. I don't think they accept PRs on GitHub. You'll probably have to chase it over the mailing list.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 10 above)
Comment (agherzan) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:39]: Indeed. This is becasue the libs are not versioned I was as suspecting. Looks good.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 2 above)
Comment (agherzan) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb]: @tewarid This doesn't seem right. First of all, _bindir_ is already there: https://git.yoctoproject.org/poky/tree/meta/conf/bitbake.conf#n327 . So those should be in by default - did you by any chance override it only with _libdir_? That would explain the error above. Same for _libdir_. It should work from defaults using SOLIBS (see the same link above).

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 3 above)
Comment (agherzan) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb]: Sounds good to me.

**Review by agherzan — CHANGES_REQUESTED** [review]

File: dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend [hunk 11]
```diff
@@ -1,2 +1,2 @@
-PACKAGECONFIG[raspberrypi] = "-Dpipelines=raspberrypi"
+PACKAGECONFIG[raspberrypi] = "-Dpipelines=raspberrypi -Dipas=raspberrypi -Dcpp_args=-Wno-unaligned-access"
```
[the remark this answers is not in the document]
Comment (tewarid) [reply dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend:1]: The `-Dcpp_args=-Wno-unaligned-access` is workaround due to https://github.com/kbingham/libcamera/pull/40 and clang 14/15.

[the remark this answers is not in the document]
Comment (agherzan) [reply dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend:1]: I see. Can we mention that in git log?

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 2 above)
Comment (agherzan) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb]: Resolved it it was dropped from the current version.

File: dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend (hunk 1 above)
Comment (tewarid) [reply dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend]: Submitted patch via email - https://lists.openembedded.org/g/openembedded-devel/message/99061.

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+EXTRA_OECMAKE:append:raspberrypi3 = " -DENABLE_COMPILE_FLAGS_FOR_TARGET=armv8-neon"
+EXTRA_OECMAKE:append:raspberrypi4 = " -DENABLE_COMPILE_FLAGS_FOR_TARGET=armv8-neon"
+
+PACKAGECONFIG[x11] = "-DENABLE_X11=1,-DENABLE_X11=0"
+PACKAGECONFIG[qt] = "-DENABLE_QT=1,-DENABLE_QT=0"
+PACKAGECONFIG[opencv] = "-DENABLE_OPENCV=1,-DENABLE_OPENCV=0"
+PACKAGECONFIG[tensorflow-lite] = "-DENABLE_TFLITE=1,-DENABLE_TFLITE=0"
+
+FILES:${PN} = "${bindir} ${libdir}"
```
Comment (agherzan) [inline recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb:37]: Was this brought back by mistake?

**Review by agherzan — CHANGES_REQUESTED** [review]

File: recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb (hunk 12 above)
Comment (tewarid) [reply recipes-multimedia/libcamera-apps/rpi-libcamera-apps_git.bb]: Yes - should be fixed now.

File: dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend (hunk 1 above)
Comment (tewarid) [reply dynamic-layers/multimedia-layer/recipes-multimedia/libcamera/libcamera.bbappend]: This is now available upstream - https://github.com/openembedded/meta-openembedded/commit/9ab32b214b54a58baf49a725dddf1c720d4195d1.

**Review by agherzan — APPROVED** [review]
Looks good. Thanks!
